src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -134,187 +135,179 @@ internal Metadata GetMetadata(string fullPath, DateTime snapshotTimestamp)
 
                 return metadata;
             }
-
-            // use temporary storage
-            var storages = new List<TemporaryStorageService.TemporaryStreamStorage>();
-            newMetadata = CreateAssemblyMetadataFromTemporaryStorage(key, storages);
-
-            // don't dispose assembly metadata since it shares module metadata
-            if (!_metadataCache.GetOrAddMetadata(key, new RecoverableMetadataValueSource(newMetadata, storages), out metadata))
+            else
             {
-                newMetadata.Dispose();
-            }
+                // use temporary storage
+                using var _ = ArrayBuilder<TemporaryStorageService.TemporaryStreamStorage>.GetInstance(out var storages);
+                var newMetadata = CreateAssemblyMetadata(key, CreateModuleMetadata);
 
-            // guarantee that the metadata is alive while we add the source to the cache
-            GC.KeepAlive(newMetadata);
+                // don't dispose assembly metadata since it shares module metadata
+                if (!_metadataCache.GetOrAddMetadata(key, new RecoverableMetadataValueSource(newMetadata, storages.ToImmutable()), out metadata))
+                {
+                    newMetadata.Dispose();
+                }
 
-            return metadata;
-        }
+                // guarantee that the metadata is alive while we add the source to the cache
+                GC.KeepAlive(newMetadata);
 
-        /// <exception cref="IOException"/>
-        /// <exception cref="BadImageFormatException" />
-        private AssemblyMetadata CreateAssemblyMetadataFromTemporaryStorage(
-            FileKey fileKey, List<TemporaryStorageService.TemporaryStreamStorage> storages)
-        {
-            var moduleMetadata = CreateModuleMetadataFromTemporaryStorage(fileKey, storages);
-            return CreateAssemblyMetadata(fileKey, moduleMetadata, storages, CreateModuleMetadataFromTemporaryStorage);
+                return metadata;
+
+                // <exception cref="IOException"/>
+                // <exception cref="BadImageFormatException" />
+                ModuleMetadata CreateModuleMetadata(FileKey moduleFileKey)
```

Comment:
>CreateModuleMetadata [](http://example.com/codeflow?start=31&length=20)

nit: using lambda instead of local function would make it easier to follow how `storages` are populated.

Reply:
will do.